File: php/tests/Unit/Domain/Rules/Event/StockMovedTest.php
<?php

declare(strict_types=1);

namespace Tests\Unit\Domain\Rules\Event;

use Alxarafe\App\Domain\Inventory\ValueObject\HandlingUnitId;
use Alxarafe\App\Domain\Rules\Event\StockMoved;
use Alxarafe\App\Domain\Rules\ValueObject\MovementType;
use Alxarafe\App\Domain\Rules\ValueObject\StockMovementId;
use Alxarafe\App\Domain\Topology\ValueObject\LocationId;
use DateTimeImmutable;
use PHPUnit\Framework\TestCase;

final class StockMovedTest extends TestCase
{
    public function testCreate(): void
    {
        $id = new StockMovementId('018e4e3a-3e7b-7b3e-8000-000000000001');
        $huId = new HandlingUnitId('018e4e3a-3e7b-7b3e-8000-000000000002');
        $from = new LocationId('018e4e3a-3e7b-7b3e-8000-000000000003');
        $to = new LocationId('018e4e3a-3e7b-7b3e-8000-000000000004');
        $now = new DateTimeImmutable();

        $event = new StockMoved($id, MovementType::TRANSFER, $huId, $from, $to, $now);

        self::assertTrue($id->equals($event->id()));
        self::assertSame(MovementType::TRANSFER, $event->type());
        self::assertTrue($huId->equals($event->huId()));
        self::assertInstanceOf(LocationId::class, $event->fromLocationId());
        self::assertInstanceOf(LocationId::class, $event->toLocationId());
        self::assertTrue($from->equals($event->fromLocationId()));
        self::assertTrue($to->equals($event->toLocationId()));
        self::assertSame($now, $event->performedAt());
    }

    public function testNullFromLocation(): void
    {
        $event = new StockMoved(
            new StockMovementId('018e4e3a-3e7b-7b3e-8000-000000000001'),
            MovementType::INBOUND,
            new HandlingUnitId('018e4e3a-3e7b-7b3e-8000-000000000002'),
            null,
            new LocationId('018e4e3a-3e7b-7b3e-8000-000000000004'),
            new DateTimeImmutable(),
        );
        self::assertNull($event->fromLocationId());
    }

    public function testNullToLocationForOutbound(): void
    {
        $event = new StockMoved(
            new StockMovementId('018e4e3a-3e7b-7b3e-8000-000000000001'),
            MovementType::OUTBOUND,
            new HandlingUnitId('018e4e3a-3e7b-7b3e-8000-000000000002'),
            new LocationId('018e4e3a-3e7b-7b3e-8000-000000000003'),
            null,
            new DateTimeImmutable(),
        );
        self::assertNotNull($event->fromLocationId());
        self::assertNull($event->toLocationId());
    }
}

File: java/src/test/java/com/alxarafe/domain/rules/event/StockMovedTest.java
package com.alxarafe.domain.rules.event;

import com.alxarafe.app.domain.inventory.valueobject.HandlingUnitId;
import com.alxarafe.app.domain.rules.event.StockMoved;
import com.alxarafe.app.domain.rules.valueobject.MovementType;
import com.alxarafe.app.domain.rules.valueobject.StockMovementId;
import com.alxarafe.app.domain.topology.valueobject.LocationId;
import org.junit.jupiter.api.Test;

import java.time.Instant;

import static org.junit.jupiter.api.Assertions.*;

class StockMovedTest {

    @Test
    void create() {
        var id = new StockMovementId("018e4e3a-3e7b-7b3e-8000-000000000001");
        var huId = new HandlingUnitId("018e4e3a-3e7b-7b3e-8000-000000000002");
        var from = new LocationId("018e4e3a-3e7b-7b3e-8000-000000000003");
        var to = new LocationId("018e4e3a-3e7b-7b3e-8000-000000000004");
        var now = Instant.now();

        var event = new StockMoved(id, MovementType.TRANSFER, huId, from, to, now);

        assertEquals(id, event.id());
        assertEquals(MovementType.TRANSFER, event.type());
        assertEquals(huId, event.huId());
        assertEquals(from, event.fromLocationId().orElseThrow());
        assertEquals(to, event.toLocationId().orElseThrow());
        assertEquals(now, event.performedAt());
    }

    @Test
    void nullFromLocation() {
        var event = new StockMoved(
                new StockMovementId("018e4e3a-3e7b-7b3e-8000-000000000001"),
                MovementType.INBOUND,
                new HandlingUnitId("018e4e3a-3e7b-7b3e-8000-000000000002"),
                null,
                new LocationId("018e4e3a-3e7b-7b3e-8000-000000000004"),
                Instant.now());
        assertTrue(event.fromLocationId().isEmpty());
    }

    @Test
    void nullToLocationForOutbound() {
        var event = new StockMoved(
                new StockMovementId("018e4e3a-3e7b-7b3e-8000-000000000001"),
                MovementType.OUTBOUND,
                new HandlingUnitId("018e4e3a-3e7b-7b3e-8000-000000000002"),
                new LocationId("018e4e3a-3e7b-7b3e-8000-000000000003"),
                null,
                Instant.now());
        assertEquals("018e4e3a-3e7b-7b3e-8000-000000000003",
                event.fromLocationId().orElseThrow().value());
        assertTrue(event.toLocationId().isEmpty());
    }
}
